src/s3/multipartcopy.rs

```diff
@@ -0,0 +1,311 @@
+        &self,
+        part: CopyUploadPart<'_>,
+    ) -> Result<CompletedPart, S3MultipartCopierError> {
+        let copy_source = format!("{}/{}", part.src.bucket(), part.src.key());
+        let copy_source_range = format!("bytes={}-{}", part.byte_range.0, part.byte_range.1);
+
+        let response = self
+            .client
+            .upload_part_copy()
+            .bucket(part.dst.bucket())
+            .key(part.dst.key())
+            .part_number(part.part_number)
+            .upload_id(part.upload_id)
+            .copy_source(copy_source)
+            .copy_source_range(copy_source_range)
+            .send()
+            .await?;
+
+        Ok(CompletedPart::builder()
+            .set_e_tag(
+                response
+                    .copy_part_result
+                    .ok_or(S3MultipartCopierError::MissingCopyPartResult)?
+                    .e_tag,
+            )
+            .part_number(part.part_number)
+            .build())
+    }
+
+    /// Completes the multipart upload by combining all parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `bucket` - The name of the destination S3 bucket.
+    /// * `key` - The key of the destination object.
+    /// * `upload_id` - The upload ID of the multipart upload.
+    /// * `parts` - A vector of completed parts.
+    ///
+    /// # Returns
+    ///
+    /// An empty result indicating success.
+    #[instrument(skip(self))]
+    async fn complete_multipart_upload(
+        &self,
+        upload_id: &str,
+        mut parts: Vec<CompletedPart>,
+    ) -> Result<(), SdkError<CompleteMultipartUploadError>> {
+        parts.sort_by_key(|part| part.part_number);
+        let completed_multipart_upload = CompletedMultipartUpload::builder()
+            .set_parts(Some(parts))
+            .build();
+
+        self.client
+            .complete_multipart_upload()
+            .bucket(self.destination.bucket())
+            .key(self.destination.key())
+            .upload_id(upload_id)
+            .multipart_upload(completed_multipart_upload)
+            .send()
+            .await?;
+
+        Ok(())
+    }
+
+    /// Performs a multipart copy of a large object from the source bucket to the destination bucket.
+    ///
+    /// # Arguments
+    ///
+    /// * `source_bucket` - The name of the source S3 bucket.
+    /// * `source_key` - The key of the source object.
+    /// * `dest_bucket` - The name of the destination S3 bucket.
+    /// * `dest_key` - The key of the destination object.
+    ///
+    /// # Returns
+    ///
+    /// An empty result indicating success.
+    #[instrument(skip(self))]
+    pub async fn send(&self) -> Result<(), S3MultipartCopierError> {
+        tracing::info!("Starting multipart copy");
+        let source_size =
+            get_source_size(&self.client, self.source.bucket(), self.source.key()).await?;
+        let part_count =
+            ((f64::value_from(source_size)? / f64::value_from(self.part_size)?).ceil()).approx()?;
```

Comment:
**suggestion:** This piece of code is begging for an edge case bug. Would be good to have explicit tests for either side of the edges to make 100% sure it does the right thing. Either as additional test cases of the full function, or factored out into a helper method with individual tests.

Reply:
Refactored this code and exercised this with property tests and type checked values.